### ISSUE TITLE
Any support for :auto type ?

### ISSUE THREAD

**Issue by Xample** [Microsoft/TypeScript#8057]
I just enabled the linter and I am facing several "expected-variable-declaration" warnings.

Those appears for example in the following case:
```ts
// Here the linter will put a warning "expected-variable-declaration"
var aNumber = 0;
// Here not (of course)
var anotherNumber:number = 0;
```

While those are warning and not errors, could we have somethign like an auto type i.e. the same as in C++, we inherit the type from the right had side affectation ? In short: a way to explicitely say we want to inherit the affected type.

This should not display any warning or error message:

```ts
var aNumber:auto = 0;
```

**Comment by kitsonk** [comment]
I don't see what that really adds and would add complexity with no value.  Without a type annotation it is implicitly "auto" or `any`.  If you don't want to have to be explicit about inferred types, then change the configuration of your linter tool.  TypeScript has `noImplicitAny` already which disallows when the type cannot be inferred.  If you want to be explicit, be explicit.

**Comment by basarat** [comment]
TypeScript official doesn't have a concept of *warnings*. Linting is offloaded to the tslint team : https://github.com/palantir/tslint 

That said since tslint doesn't have access to *typechecker* it will probably not exist there either for a while :rose: 

PS: I'll definitely add it as a *quick fix* option in `alm` at some point

**Comment by sandersn** [comment]
As @basarat says, this is a tslint problem. TypeScript's syntax for C++'s `auto` is exactly `var aNumber = 0;`, plus `--noImplicitAny` to issue an error when a type can't be inferred.

The solution to your problem appears to be:

1. Decide whether to turn off the rule. If you would use `auto` in C++ in the same situation then I think this is the right option.
2. Do as the rule asks and annotate every declaration. If you dislike `auto`, then this is the right option. But it's a not a good option for JavaScript in my opinion.

Also, can you link to the source of the rule? I couldn't find it when searching tslint's repo.
